### PR TITLE
Core: Common metadata for TableMetadata and ViewMetadata

### PR DESCRIPTION
**Pull request by nk1506**
Ref Issue #9514 

[TableMetadata](https://github.com/apache/iceberg/blob/3b8fd904ea6c3b0f81afdae59b17ab12e45c5faf/core/src/main/java/org/apache/iceberg/TableMetadata.java)  and [ViewMetadata](https://github.com/apache/iceberg/blob/3b8fd904ea6c3b0f81afdae59b17ab12e45c5faf/core/src/main/java/org/apache/iceberg/view/ViewMetadata.java) have some common methods , Which have been abstracted out to IcebergMetadata now. It should help to have common code for Iceberg Table and View implementations with different catalog.

### CODE REVIEW

**Comment by nk1506** [comment]
cc: @szehon-ho , @nastra , @pvary

**Comment by nastra** [comment]
Sorry for the delay. I have some feedback and I'll take a closer look at this (and the Hive + View support PR) once 1.5.0 is out

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:69]: Question, can we put Override and Value.parameter here, and remove it from TableMetadata?  It will be more clear.

File: core/src/main/java/org/apache/iceberg/BaseMetadata.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseMetadata.java:29]: typo for ane.   I would actually may be renmove the first sentence, it looks like it does not add that much more info than the first sentence.  So just

```
A base class for {@link TableMetadata} and {@link org.apache.iceberg.view.ViewMetadata}.
```

**Review by szehon-ho — COMMENTED** [review]
Thanks, it looks better.  Left some comment.

For reference, we already had some existing discussion on this PR on : https://github.com/nk1506/iceberg/pull/9

File: core/src/main/java/org/apache/iceberg/BaseMetadata.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseMetadata.java:44]: Is this necessary to make this in the common base class, to allow ViewOperations  ?   Should we remove it , if it is not required?

Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:44]: Yes , Schema is needed for Both table and view operations. Ref: https://github.com/apache/iceberg/pull/8907/files#diff-db46657b84d66e084e15f31b8dab21577efb2ae7102863f94c6c9477782de676R160

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java
Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:69]: Could you please add more context? i couldn't understand this part.

File: core/src/main/java/org/apache/iceberg/BaseMetadata.java
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: Just one question. Is this refactoring enough? 

If we Introduce `BaseMetadata`, I expect the `TableOperations` and `ViewOperations` to have a base class called `BaseOperations` which uses this `BaseMetadata` for interfaces like 
`void commit(BaseMetadata base, BaseMetadata metadata);`

Without that I don't see much value in this refactoring.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: may be we can have an end to end refactoring DRAFT PR and take a decision based on that.

Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: I agree, But i think those changes we can do later once this interface pattern gets approved. Or may be with this patch itself. 
WDYT ?

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: we can have a separate draft patch. I think doing that refactoring can help us know whether the current interfaces in `BaseMetadata` is enough or do we need more changes here.

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:69]: Yea exactly, will take be able to remove @Value.Parameter in the BaseMetadata, and keep the annotation all in the ViewMetadata.

File: core/src/main/java/org/apache/iceberg/BaseMetadata.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: Yea I think the main refactor is over at https://github.com/apache/iceberg/pull/8907 (including baseOperations and everything).  I do agree that it is a bit hard to read, I have to search over there to find what is needed and not, and so may miss something :).

Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:31]: Actually sorry the refactor of the Operations is split in https://github.com/apache/iceberg/pull/9461, #8907 is the whole patch.

Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/BaseMetadata.java:27]: This is draft [PR](https://github.com/nk1506/iceberg/pull/10/files) . with https://github.com/apache/iceberg/pull/9461 and https://github.com/apache/iceberg/pull/9682 .

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java]: @nk1506 let me know if the question makes more sense.

Comment (nk1506) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java]: @szehon-ho , Since we are planning to keep common members with ViewMetadata too with `@Override`. Do we have a need of keeping `@Value.Parameter` in ViewMetadata. I think only override should works. 
WDYT ? I have updated the patch with only `@Override` and removed `@Value.Parameter` from everywhere.

File: core/src/main/java/org/apache/iceberg/BaseMetadata.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseMetadata.java:33]: Slightly prefer not to have it, as most Iceberg classes don't have it.  But understand its from the old class.

**Review by szehon-ho — APPROVED** [review]
This looks ok to me as a refactor.  I wonder if we need more input from @ajantha-bhat or @nastra .  It is a bit hard to follow without https://github.com/nk1506/iceberg/pull/10/files

**Comment by nastra** [comment]
let's hold off on this one until 1.5.0 is out and the Hive view PR is reviewed so that we can better decide in which direction we'd want to go

**Comment by szehon-ho** [comment]
Can we please continue this effort , it has been stalled 2 weeks?  

Let's have a PR that is the end to end version (please not on your own github, but against Iceberg github).

**Comment by nk1506** [comment]
Thanks @szehon-ho for looking into this again. I will raise another combined PR with [CommonMetadata](https://github.com/apache/iceberg/pull/9682), [CommonHiveOperation](https://github.com/apache/iceberg/pull/9461) and [Hive-View](https://github.com/apache/iceberg/pull/8907)

**Comment by nk1506** [comment]
> Can we please continue this effort , it has been stalled 2 weeks?
> 
> Let's have a PR that is the end to end version (please not on your own github, but against Iceberg github).

Hi @szehon-ho , I have created a new PR here https://github.com/apache/iceberg/pull/9852 .

**Comment by nastra** [comment]
I reviewed https://github.com/apache/iceberg/pull/9852/ and also took a look at the metadata abstraction but I'm not convinced it actually makes stuff easier. Most of the time certain Hive methods only require 1-2 params (properties, schema, ...) from table/view metadata, so passing those directly makes more sense. 

Additionally, this abstraction can actually hide errors, where we check for certain table properties by providing `BaseMetadata`. These properties don't exist for views, so what will happen if we check `TableProperties.ENGINE_HIVE_ENABLED` on a table and on view?

**Comment by szehon-ho** [comment]
Thanks, i commented on the other thread in https://github.com/apache/iceberg/pull/9852#discussion_r1516373990 the original reason (saw that one first)

Can you clarify this concern?

> Additionally, this abstraction can actually hide errors, where we check for certain table properties by providing BaseMetadata. These properties don't exist for views, so what will happen if we check TableProperties.ENGINE_HIVE_ENABLED on a table and on view?

This is just a class hiearchy.  The view now gets methods like String property(String property, String defaultValue), which is just shorthand for viewMetadata.getProperties().getOrDefault(property, defaultValue).  Strictly speaking, its not necssary.

**Comment by nk1506** [comment]
Common code abstraction on Hive-Table/View has been parked for later. BaseMetadata for Table and View might not be required. We can fulfil the same using different strategy.

**Pull request closed without merging**
